File: e2e/pages/login.page.ts
/**
 * login.page.ts
 *
 * Page Object for /login.
 *
 * Selector strategy:
 *  - Inputs   → getByPlaceholder (resilient to class/ID churn)
 *  - Buttons  → getByRole with name (accessible, spec-stable)
 *  - Errors   → semantic CSS classes already used by the app (.bg-red-50)
 */
import { Page, Locator, expect } from '@playwright/test';
import { BasePage } from './base.page';

export class LoginPage extends BasePage {
  readonly path = '/login';

  // ── Locators ─────────────────────────────────────────────────────────────
  readonly emailInput:    Locator;
  readonly passwordInput: Locator;
  readonly submitButton:  Locator;
  readonly globalError:   Locator;
  readonly fieldErrors:   Locator;
  readonly registerLink:  Locator;
  readonly forgotLink:    Locator;
  readonly googleButton:  Locator;
  readonly githubButton:  Locator;

  constructor(page: Page) {
    super(page);
    this.emailInput    = page.getByPlaceholder('name@example.com');
    this.passwordInput = page.getByPlaceholder('••••••••');
    this.submitButton  = page.getByRole('button', { name: 'Sign In Now' });
    this.globalError   = page.locator('.bg-red-50').first();
    this.fieldErrors   = page.locator('p.text-red-600, p.text-red-500');
    this.registerLink  = page.getByRole('link', { name: /register here/i });
    this.forgotLink    = page.getByRole('link', { name: /recovery options/i });
    this.googleButton  = page.getByRole('button', { name: /google/i });
    this.githubButton  = page.getByRole('button', { name: /github/i });
  }

  // ── Actions ──────────────────────────────────────────────────────────────

  async open(): Promise<this> {
    await this.goto(this.path);
    await expect(this.emailInput).toBeVisible();
    return this;
  }

  async fillEmail(email: string): Promise<this> {
    await this.emailInput.fill(email);
    return this;
  }

  async fillPassword(password: string): Promise<this> {
    await this.passwordInput.fill(password);
    return this;
  }

  async submit(): Promise<this> {
    await this.submitButton.click();
    return this;
  }

  /** High-level login that waits for redirect on success. */
  async login(email: string, password: string): Promise<void> {
    await this.fillEmail(email);
    await this.fillPassword(password);
    await this.submit();
  }

  /** Expect a visible global (server-side) error banner. */
  async expectGlobalError(text?: string | RegExp): Promise<void> {
    await expect(this.globalError).toBeVisible({ timeout: 8_000 });
    if (text) await expect(this.globalError).toContainText(text);
  }

  /** Expect at least one inline field-level error message. */
  async expectFieldError(): Promise<void> {
    const visible = this.fieldErrors.first();
    await expect(visible).toBeVisible({ timeout: 5_000 });
  }

  async expectRedirectAway(): Promise<void> {
    await this.page.waitForURL((url) => !url.pathname.includes('/login'), {
      timeout: 15_000,
    });
  }
}

File: selenium/pages/login_page.py
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC


class LoginPage:
    PATH = "/login"

    # Selectors
    EMAIL_INPUT    = (By.CSS_SELECTOR, "input[type='email']")
    PASSWORD_INPUT = (By.CSS_SELECTOR, "input[type='password']")
    SUBMIT_BTN     = (By.CSS_SELECTOR, "button[type='submit']")
    GLOBAL_ERROR   = (By.CSS_SELECTOR, "div.bg-red-50")
    FIELD_ERROR    = (By.CSS_SELECTOR, "p.text-red-600, p.text-red-500")
    REGISTER_LINK  = (By.CSS_SELECTOR, "a[href*='register']")

    def __init__(self, driver, base_url, timeout=15):
        self.driver = driver
        self.base_url = base_url
        self.wait = WebDriverWait(driver, timeout)

    def open(self):
        self.driver.get(f"{self.base_url}{self.PATH}")
        self.wait.until(EC.visibility_of_element_located(self.EMAIL_INPUT))
        return self

    def fill_email(self, email):
        field = self.wait.until(EC.visibility_of_element_located(self.EMAIL_INPUT))
        field.clear()
        field.send_keys(email)
        return self

    def fill_password(self, password):
        field = self.wait.until(EC.visibility_of_element_located(self.PASSWORD_INPUT))
        field.clear()
        field.send_keys(password)
        return self

    def submit(self):
        self.wait.until(EC.element_to_be_clickable(self.SUBMIT_BTN)).click()
        return self

    def login(self, email, password):
        self.fill_email(email)
        self.fill_password(password)
        self.submit()
        return self

    def wait_for_redirect_away(self):
        self.wait.until(lambda d: self.PATH not in d.current_url)

    def get_global_error(self):
        return self.wait.until(EC.visibility_of_element_located(self.GLOBAL_ERROR))

    def get_field_errors(self):
        return self.wait.until(EC.presence_of_all_elements_located(self.FIELD_ERROR))

    def click_register_link(self):
        self.wait.until(EC.element_to_be_clickable(self.REGISTER_LINK)).click()
